src/spec.rs

```diff
@@ -120,11 +120,20 @@ pub(crate) fn ka_orchard(sk: &pallas::Scalar, b: &pallas::Point) -> pallas::Poin
 ///
 /// [concreteextractorpallas]: https://zips.z.cash/protocol/nu5.pdf#concreteextractorpallas
 pub(crate) fn extract_p(point: &pallas::Point) -> pallas::Base {
-    if let Some((x, _)) = point.to_affine().get_xy().into() {
-        x
-    } else {
-        pallas::Base::zero()
-    }
+    point
+        .to_affine()
+        .coordinates()
+        .map(|c| *c.x())
+        .unwrap_or_else(pallas::Base::zero)
+}
+
+/// Coordinate extractor for Pallas.
+///
+/// Defined in [Zcash Protocol Spec § 5.4.9.7: Coordinate Extractor for Pallas][concreteextractorpallas].
+///
+/// [concreteextractorpallas]: https://zips.z.cash/protocol/nu5.pdf#concreteextractorpallas
+pub(crate) fn extract_p_bottom(point: CtOption<pallas::Point>) -> CtOption<pallas::Base> {
```

Comment:
Should `IncompletePoint` be moved out of sinsemilla::addition and `extract_p_bottom` defined over that newtype?

Reply:
I decided against that because incomplete addition is an implementation detail of Sinsemilla; what matters externally is the return type. The function signature `P u {bottom} -> P::Base u {bottom}` maps more intuitively to `Option<P> -> Option<P::Base>` with `bottom` represented as `None`.

Reply:
Ordinarily, it's a design smell to take an `Option` as input unless you're doing to do something other than just map over it, because the caller is constrained by the types to do so anyway. It's not strictly wrong, of course, but perhaps just superfluous? If this is a pure function, then there's no possible implementation other than the one given, and the caller could do that just as easily?

Reply:
This directly corresponds to a function in the spec, and I think it's better to make that correspondence clear at call sites.

Reply:
@nuttycom you're correct that the caller could do this map themselves. The existence of `ectract_p_bottom` as a function is purely for the benefit of people reviewing against the spec.